Crash in MethodCallExpr with Reference to TypeParameter

### Code

```rust
extern "C" {
    fn printf(s: *const i8, ...);
}

struct Foo(i32);
trait Bar {
    fn baz(&self);
}

impl Bar for Foo {
    fn baz(&self) {
        unsafe {
            let a = "%i\n\0";
            let b = a as *const str;
            let c = b as *const i8;

            printf(c, self.0);
        }
    }
}

fn type_bound<T: Bar>(t: &T) {
    t.baz();
}

fn main() {
    let a = &Foo(123);

    type_bound(a);
}

```

### Meta

- What version of Rust GCC were you using, git sha if possible. 4493f1c927c79e5b0933d26e9597044790c22f12

### Error output

```
rust1: note: END POST-EXPANSION AST DUMP
test.rs:7:12: warning: unused name 'self'
    7 |     fn baz(&self);
      |            ^
test.rs:11:5: warning: unused name '<Foo as Bar>::baz'
   11 |     fn baz(&self) {
      |     ^
rust1: internal compiler error: in visit, at rust/backend/rust-compile.cc:187
```

<!--
Include a backtrace if available.
-->
<details><summary>Backtrace</summary>
  <p>
  
  ```
rust1: internal compiler error: in visit, at rust/backend/rust-compile.cc:187
0x10181bd Rust::Compile::CompileExpr::visit(Rust::HIR::MethodCallExpr&)
        ../../gccrs/gcc/rust/backend/rust-compile.cc:187
0x10754fb Rust::HIR::MethodCallExpr::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:3927
0x1021424 Rust::Compile::CompileExpr::Compile(Rust::HIR::Expr*, Rust::Compile::Context*)
        ../../gccrs/gcc/rust/backend/rust-compile-expr.h:39
0x10766d1 Rust::HIR::ExprStmtWithoutBlock::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:4425
0x1019717 Rust::Compile::HIRCompileBase::compile_function_body(Bfunction*, std::unique_ptr<Rust::HIR::BlockExpr, std::default_delete<Rust::HIR::BlockExpr> >&, bool)
        ../../gccrs/gcc/rust/backend/rust-compile.cc:493
0x1075d83 Rust::HIR::Function::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:4173
0x104c7a7 Rust::Compile::ResolvePathRef::query_compile(unsigned int, Rust::TyTy::BaseType*, Rust::HIR::PathIdentSegment const&, Rust::Analysis::NodeMapping const&, Location, bool)
        ../../gccrs/gcc/rust/backend/rust-compile-resolve-path.cc:122
0x104c63c Rust::Compile::ResolvePathRef::resolve(Rust::HIR::PathIdentSegment const&, Rust::Analysis::NodeMapping const&, Location, bool)
        ../../gccrs/gcc/rust/backend/rust-compile-resolve-path.cc:103
0x104c2d2 Rust::Compile::ResolvePathRef::visit(Rust::HIR::PathInExpression&)
        ../../gccrs/gcc/rust/backend/rust-compile-resolve-path.cc:39
0x1074e39 Rust::HIR::PathInExpression::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:3741
0x1074e39 Rust::HIR::PathInExpression::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:3741
0x1021424 Rust::Compile::CompileExpr::Compile(Rust::HIR::Expr*, Rust::Compile::Context*)
        ../../gccrs/gcc/rust/backend/rust-compile-expr.h:39
0x1017a85 Rust::Compile::CompileExpr::visit(Rust::HIR::CallExpr&)
        ../../gccrs/gcc/rust/backend/rust-compile.cc:89
0x10754c3 Rust::HIR::CallExpr::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:3921
0x1021424 Rust::Compile::CompileExpr::Compile(Rust::HIR::Expr*, Rust::Compile::Context*)
        ../../gccrs/gcc/rust/backend/rust-compile-expr.h:39
0x10766d1 Rust::HIR::ExprStmtWithoutBlock::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:4425
0x1019717 Rust::Compile::HIRCompileBase::compile_function_body(Bfunction*, std::unique_ptr<Rust::HIR::BlockExpr, std::default_delete<Rust::HIR::BlockExpr> >&, bool)
        ../../gccrs/gcc/rust/backend/rust-compile.cc:493
0x1075d83 Rust::HIR::Function::accept_vis(Rust::HIR::HIRVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:4173
0x1017751 Rust::Compile::CompileCrate::go()
        ../../gccrs/gcc/rust/backend/rust-compile.cc:50
0x10175d1 Rust::Compile::CompileCrate::Compile(Rust::HIR::Crate&, Rust::Compile::Context*)
        ../../gccrs/gcc/rust/backend/rust-compile.cc:40
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
[Inferior 1 (process 300771) exited with code 04]
  ```
  
  </p>
</details>

--- a/gcc/rust/backend/rust-compile.cc
+++ b/gcc/rust/backend/rust-compile.cc
@@ -22,6 +22,7 @@
 #include "rust-compile-struct-field-expr.h"
 #include "rust-hir-trait-resolve.h"
 #include "rust-hir-path-probe.h"
+#include "rust-hir-dot-operator.h"
 
 namespace Rust {
 namespace Compile {
@@ -171,10 +172,11 @@ CompileExpr::visit (HIR::MethodCallExpr &expr)
 	  // item so its up to us to figure out if this path should resolve
 	  // to an trait-impl-block-item or if it can be defaulted to the
 	  // trait-impl-item's definition
+
+	  auto root = receiver->get_root ();
 	  std::vector<Resolver::PathProbeCandidate> candidates
 	    = Resolver::PathProbeType::Probe (
-	      receiver, expr.get_method_name ().get_segment (), true, false,
-	      true);
+	      root, expr.get_method_name ().get_segment (), true, false, true);
 
 	  if (candidates.size () == 0)
 	    {
@@ -185,6 +187,11 @@ CompileExpr::visit (HIR::MethodCallExpr &expr)
 							  &trait_item_ref);
 	      rust_assert (ok);				    // found
 	      rust_assert (trait_item_ref->is_optional ()); // has definition
+
+	      // FIXME Optional means it has a definition and an associated
+	      // block which can be a default implementation, if it does not
+	      // contain an implementation we should actually return
+	      // error_mark_node
 
 	      TyTy::BaseType *self_type = nullptr;
 	      if (!ctx->get_tyctx ()->lookup_type (
@@ -209,10 +216,19 @@ CompileExpr::visit (HIR::MethodCallExpr &expr)
 	    }
 	  else
 	    {
-	      Resolver::PathProbeCandidate &candidate = candidates.at (0);
-	      rust_assert (candidate.is_impl_candidate ());
+	      std::vector<Resolver::Adjustment> adjustments;
+	      Resolver::PathProbeCandidate *candidate
+		= Resolver::MethodResolution::Select (candidates, root,
+						      adjustments);
 
-	      HIR::ImplItem *impl_item = candidate.item.impl.impl_item;
+	      // FIXME this will be a case to return error_mark_node, there is
+	      // an error scenario where a Trait Foo has a method Bar, but this
+	      // receiver does not implement this trait or has an incompatible
+	      // implementation and we should just return error_mark_node
+	      rust_assert (candidate != nullptr);
+	      rust_assert (candidate->is_impl_candidate ());
+
+	      HIR::ImplItem *impl_item = candidate->item.impl.impl_item;
 
 	      TyTy::BaseType *self_type = nullptr;
 	      if (!ctx->get_tyctx ()->lookup_type (

--- a/gcc/testsuite/rust/execute/torture/trait4.rs
+++ b/gcc/testsuite/rust/execute/torture/trait4.rs
@@ -1,0 +1,36 @@
+/* { dg-output "123\n" }*/
+extern "C" {
+    fn printf(s: *const i8, ...);
+}
+
+struct Foo(i32);
+trait Bar {
+    fn baz(&self);
+    // { dg-warning "unused name" "" { target *-*-* } .-1 }
+}
+
+impl Bar for Foo {
+    fn baz(&self) {
+        // { dg-warning "unused name" "" { target *-*-* } .-1 }
+        unsafe {
+            let a = "%i\n\0";
+            let b = a as *const str;
+            let c = b as *const i8;
+
+            printf(c, self.0);
+        }
+    }
+}
+
+fn type_bound<T: Bar>(t: &T) {
+    t.baz();
+}
+
+fn main() -> i32 {
+    let a;
+
+    a = &Foo(123);
+    type_bound(a);
+
+    0
+}

